Uniformise documentation of _opt functions in Pervasive

mdesharnais opened the pull request:
Use the format "Same as [f], but returns [None] instead of raising."

nojb commented:
A tiny nitpick, but I would write `return` rather than `returns` for uniformity with the rest of the doc strings.
Did you check the rest of the stdlib ?

mdesharnais commented:
> A tiny nitpick, but I would write return rather than returns for uniformity with the rest of the doc strings.

Ok

> Did you check the rest of the stdlib ?

No, I just noticed a typo in `read_int_opt` and wanted to correct it. Than I notices those two functions having a different description style. If it is wished, I can go through every mli file and uniformise the hole stdlib.

xavierleroy commented:
"Return _xxx_" is an imperative mood while "Same as _f_ but [it] returns _xxx_" is an indicative mood.  

My English is not good enough to tell you which form to use.  Please check with native English speakers who know grammar before "uniformising" spelling.

alainfrisch commented:
I think we want to encourage the use of option-returning raising variants (and perhaps, some day, deprecate the exception-raising ones).  As a weak signal, if one is ready to spend some energy on the stdlib documentation, I'd prefer to fully document the option-returning variants, and refer to them for the documentation of the exception-raising ones.  At least, let's avoid doing the opposite.

mdesharnais commented:
> Please check with native English speakers who know grammar before "uniformising" spelling.

I am sorry if I was not clear enough. I noticed the documentation of `read_int_opt` is "Same as [read_int_opt], but returns [None] instead of raising." But this is a recursive definition and I assumed that `read_int` is meant.

While removing the `_opt`, I noticed that the documentation of `bool_of_string_opt` and `read_float_opt` are repeating what those functions do instead of pointing the difference with their throwing counterpart. I personally think it is better to be uniform in the wording and just copied/pasted the existing sentence from `read_int_opt` and change the referenced function name.

Except that the documentation of `read_float_opt` does not say the function throws an exception in case of invalid input: it just says the result is undefined. So I tried to minimally change the wording to reflect it.

Now that we are so explicit, is there a reason the fact that the function throws an exception is not documented?

mdesharnais commented:
> I think we want to encourage the use of option-returning raising variants (and perhaps, some day, deprecate the exception-raising ones).

I am glad to read it!

> As a weak signal, if one is ready to spend some energy on the stdlib documentation, I'd prefer to fully document the option-returning variants, and refer to them for the documentation of the exception-raising ones. At least, let's avoid doing the opposite.

I agree with you, if this is really the wanted direction, I can reverse the references so that the throwing version references the option returning version.

mdesharnais commented:
Hello,
I updated the commit with the suggestions of @alainfrisch (i.e. the raising functions refer to the `option`-returning ones) and @nojb and @xavierleroy (i.e. use imperative instead of indicative mood).

Is it a welcomed change and if yes, should I proceed with the rest of the documentation?

Also, is there a reason for `read_float_opt` referring to an unspecified result in case of an invalid floating-point number? Should the description be changed to refer to the raised exception?

alainfrisch commented:
> Is it a welcomed change 

I support it, but please give some time for others (@xavierleroy?) to complain, so that you don't loose time if there is no consensus.

Suggested improvement, if the change is accepted in principle: Can you change the references to use the syntax `{!...}`, in order to get clickable links in the generated doc?  Also perhaps move the `_opt` function before their raising counterparts in the .mli, to avoid forward references.

> is there a reason for read_float_opt referring to an unspecified result in case of an invalid floating-point number?

I don't see any reason, considering the implementation relies on `float_of_string`, which specifies the behavior.

(Not related to your PR, but I realized that `read_*` don't say anything about the `End_of_file` condition.  We should at least document that as well. And perhaps  add `*_opt` variants for `input_*` and `read_line` (and arrange so that `read_int` and friends also return `None` at eof).)

mdesharnais commented:
Hello, is there some new comment/input on this subject?

mdesharnais commented:
Happy new year! Is there hope this pull request will be eventually merged?

damiendoligez reviewed (changes requested):
This will be good to go after you change the standard phrase to "Same as {!f}, but raise [Exn] instead of returning [None]." as requested by @alainfrisch.

mdesharnais commented:
True, I somehow missed to put the link. I also add a separate commit to document the exception raised by `read_float` on invalid input.

damiendoligez commented:
@authchir could you rebase? Then I'll merge.

mdesharnais committed:
Uniformise documentation of `_opt` functions in Pervasives

The `option`-returning versions now document the behaviour of the
functions and are refered to by the raising versions using the following
pattern:

> Same as {!Pervasives.f_opt}, but raise [x] instead of returning [None].

mdesharnais committed:
Document function `read_float` raising an exception

mdesharnais commented:
@damiendoligez Here it is. Sorry for the delay.

damiendoligez commented:
The travis failures are unrelated to this PR (they are fixed by 7a0c95c7ae03f549b5135fdfee184ba9d6733a9c).
